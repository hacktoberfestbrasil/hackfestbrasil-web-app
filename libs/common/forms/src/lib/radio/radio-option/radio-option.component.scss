:host {
  margin: 4px;
  label {
    display: inline-block;
    position: relative;
    font-size: 40px;
    line-height: 50px;
    padding: 0 0 0 80px;
    vertical-align: top;
    color: rgba(255, 255, 255, 0.8);
    cursor: pointer;
    transition: color 0.3s ease;
  }
  input[type='radio'],
  label::before {
    width: 50px;
    height: 50px;
    top: 50%;
    left: 0;
    margin-top: -25px;
    position: absolute;
    cursor: pointer;
    box-sizing: border-box;
  }

  input[type='radio'] {
    opacity: 0;
    appearance: none;
    display: inline-block;
    vertical-align: middle;
    z-index: 100;
  }

  label::before {
    content: '';
    border: 4px solid #93C2DB;
    transition: opacity 0.3s ease;
    border-radius: 50%;
  }

  input[type='radio']:checked + label {
    color: #fff;
  }

  input[type='radio']:checked + label::before {
    border: 4px solid #ffffff;
  }

  /* General SVG and path styles */
  svg {
    position: absolute;
    width: 40px;
    height: 40px;
    top: 50%;
    margin-top: -20px;
    left: 5px;
    pointer-events: none;
  }

  svg path {
    stroke: rgba(255, 138, 226, 1);
    stroke-width: 13px;
    stroke-linecap: round;
    stroke-linejoin: round;
    fill: none;
    transition: 0.8s stroke-dasharray ease, 0.8s stroke-dashoffset ease,
      0.8s stroke ease;
  }

  /* Radio */
  .hackfest-radio input[type='radio'] + label svg path {
    stroke-dasharray: 490px 600px;
    stroke-dashoffset: 491px;
    transition: 0.8s stroke-dasharray ease, 0.8s stroke-dashoffset ease,
      0.8s stroke ease;
  }

  .hackfest-radio input[type='radio']:checked + label svg path {
    stroke-dashoffset: 0px;
  }
}
