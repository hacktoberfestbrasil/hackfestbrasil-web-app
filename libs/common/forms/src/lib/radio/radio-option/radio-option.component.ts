import { EventInputTarget } from './../../types/events.type';
import {
  ChangeDetectionStrategy,
  Component,
  ElementRef,
  EventEmitter,
  Input,
  Output,
  ViewChild,
} from '@angular/core';

let nextId = 0;

@Component({
  selector: 'hackfest-radio-option',
  templateUrl: './radio-option.component.html',
  styleUrls: ['./radio-option.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class HackfestRadioOptionComponent {
  @ViewChild('input', { static: true }) _el: ElementRef<HTMLInputElement>;
  get el() {
    return this._el.nativeElement;
  }

  name: string;

  public _value: any;
  public get value(): any {
    return this._value;
  }

  @Input()
  public set value(value: any) {
    this._value = value;
  }

  @Input()
  public set id(value: string) {
    this._id = value;
  }
  public get id(): string {
    return this._id;
  }
  private _id = `hackfest-radio-option-${nextId++}`;

  @Output()
  checkedChange = new EventEmitter<HackfestRadioOptionComponent>();

  onChangeEvent() {
    this.checkedChange.emit(this);
  }
}
