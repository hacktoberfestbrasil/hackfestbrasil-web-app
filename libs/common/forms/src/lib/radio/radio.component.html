<div class="hackfest-radio">
  <ng-content select="hackfest-radio-option"></ng-content>
</div>
