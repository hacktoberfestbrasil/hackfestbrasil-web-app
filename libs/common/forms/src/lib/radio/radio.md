## Import

```ts
import { HackfestRadioModule } from '@hackfest/forms';
```


