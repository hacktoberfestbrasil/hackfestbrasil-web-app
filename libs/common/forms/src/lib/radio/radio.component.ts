import {
  Self,
  Input,
  Output,
  Optional,
  ViewChild,
  Component,
  Injectable,
  ElementRef,
  forwardRef,
  OnDestroy,
  EventEmitter,
  AfterViewInit,
  AfterContentInit,
  ChangeDetectionStrategy,
  ContentChildren,
  QueryList,
} from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { EventInputTarget } from './../types';
import {
  AbstractControl,
  FormControl,
  NgControl,
  NG_VALUE_ACCESSOR,
} from '@angular/forms';
import { HackfestControlAccessor } from '../abstractions';
import { HackfestRadioOptionComponent } from './radio-option';

let nextId = 0;

@Injectable()
export class HackfestRadio extends HackfestControlAccessor {
  @Input()
  public set disabled(value: boolean) {
    this._disabled = value;
  }
  @Input()
  public set value(value: any) {
    this._value = value;
  }
}

const HackfestRadioProvider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => HackfestRadio),
  multi: true,
};

@Component({
  selector: 'hackfest-radio',
  templateUrl: './radio.component.html',
  styleUrls: ['./radio.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [HackfestRadio, HackfestRadioProvider],
})
export class HackfestRadioComponent extends HackfestRadio
  implements AfterContentInit, AfterViewInit, OnDestroy {
  destroy$ = new Subject<void>();

  @ContentChildren(forwardRef(() => HackfestRadioOptionComponent), {
    descendants: true,
  })
  _radios: QueryList<HackfestRadioOptionComponent>;

  @ViewChild('input', { static: true }) _el: ElementRef<HTMLInputElement>;
  get el() {
    return this._el.nativeElement;
  }

  @Input()
  public set id(value: string) {
    this._id = value;
  }
  public get id(): string {
    return this._id;
  }
  private _id = `hackfest-radio-${nextId++}`;

  @Output()
  valueChange: EventEmitter<any> = new EventEmitter<any>();

  @Output()
  checkedChange = new EventEmitter<boolean>(false);

  control: AbstractControl;

  constructor(@Optional() @Self() public ngControl: NgControl) {
    super();
  }

  ngAfterContentInit() {
    this.control = this.ngControl?.control
      ? this.ngControl.control
      : new FormControl();
  }
  ngAfterViewInit() {
    this.ngControl?.control?.valueChanges
      .pipe(takeUntil(this.destroy$))
      .subscribe((value) => this.valueChange.emit(value));
    this._radios.map((item) => this.onChangeOption(item));
  }
  onChangeOption(option: HackfestRadioOptionComponent) {
    option.name = this.id;
    option.checkedChange
      .pipe(takeUntil(this.destroy$))
      .subscribe(({ value, el }) => {
        this.control.setValue(value);
        this.checkedChange.emit(el.checked);
        this.valueChange.emit(value);
      });
  }
  onChangeEvent(evt: EventInputTarget) {
    this.onChange(evt.target.value);
    this.checkedChange.emit(evt.target.checked);
  }

  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }
}
