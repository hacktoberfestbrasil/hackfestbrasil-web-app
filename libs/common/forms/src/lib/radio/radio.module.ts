import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HackfestRadioComponent } from './radio.component';
import { HackfestRadioOptionComponent } from './radio-option/radio-option.component';

export * from './radio.component';
export * from './radio-option';

@NgModule({
  imports: [FormsModule, CommonModule, ReactiveFormsModule],
  declarations: [HackfestRadioComponent, HackfestRadioOptionComponent],
  exports: [HackfestRadioComponent, HackfestRadioOptionComponent],
})
export class HackfestRadioModule {}
