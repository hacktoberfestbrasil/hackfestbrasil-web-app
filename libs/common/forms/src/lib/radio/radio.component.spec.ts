import { FormControl, FormsModule, ReactiveFormsModule } from '@angular/forms';
import { Component } from '@angular/core';
import {
  Spectator,
  SpectatorHost,
  createHostFactory,
  createComponentFactory,
} from '@ngneat/spectator';


import { HackfestRadioComponent } from './radio.component';
import { HackfestRadioOptionComponent } from './radio-option';

const VIEWS = {
  onlyOne: `
    <hackfest-radio [formControl]="form">
      <hackfest-radio-option value="1">1</hackfest-radio-option>
    </hackfest-radio>
  `,
  withOptions: `
    <hackfest-radio [formControl]="form">
      <hackfest-radio-option *ngFor="let c of choices" [value]="c.k">{{ c.v }}</hackfest-radio-option>
    </hackfest-radio>
  `,
  withCustomId: `
    <hackfest-radio-option id="x" value="1">1</hackfest-radio-option>
  `,
};

const OPTIONS = Array.from(['HTML', 'CSS']);

@Component({ selector: 'hackfest-custom-form', template: '' })
class CustomFormComponent {
  form = new FormControl();
  choices = OPTIONS;
}

describe('HackfestRadioComponent', () => {
  describe('Host', () => {
    let spectator: SpectatorHost<HackfestRadioComponent, CustomFormComponent>;
    const createHost = createHostFactory({
      component: HackfestRadioComponent,
      host: CustomFormComponent,
      declarations: [HackfestRadioOptionComponent],
      imports: [ReactiveFormsModule],
    });

    describe('Views', () => {
      it('should display only one option', () => {
        spectator = createHost(VIEWS.onlyOne);
        expect(spectator.query('.hackfest-radio')).toMatchSnapshot();
      });

      it('should display radio with options', () => {
        spectator = createHost(VIEWS.withOptions);
        expect(spectator.query('.hackfest-radio')).toMatchSnapshot();
      });
    });

    describe('Props', () => {
      it('should display only one option', () => {
        spectator = createHost(VIEWS.onlyOne);
        const input = spectator.queryAll('hackfest-radio-option');
        expect(input.length).toEqual(1);
      });

      it('should create an instance with 2 options', () => {
        spectator = createHost(VIEWS.withOptions);
        const input = spectator.queryAll('hackfest-radio-option');
        expect(input.length).toEqual(2);
      });
    });
  });

  describe('Forms', () => {
    let spectator: Spectator<HackfestRadioComponent>;
    const createComponent = createComponentFactory({
      component: HackfestRadioComponent,
      imports: [FormsModule, ReactiveFormsModule],
    });

    it('should create', () => {
      spectator = createComponent();
      expect(spectator.component).toBeTruthy();
    });

    it('should value changed if set value', () => {
      spectator.component.control.setValue('abc');
      expect(spectator.component.control.value).toBe('abc');
    });

    it('should value changed if set value', async () => {
      spyOn(spectator.component.valueChange, 'emit');
      spectator.component.control.valueChanges.subscribe(() => {
        expect(spectator.component.onChangeEvent).toBeCalledTimes(1);
        expect(spectator.component.valueChange.emit).toBeCalledTimes(1);
        expect(spectator.component.checkedChange.emit).toBeCalledTimes(1);
      });
      spectator.component.control.setValue('qwe');
    });
  });
});

