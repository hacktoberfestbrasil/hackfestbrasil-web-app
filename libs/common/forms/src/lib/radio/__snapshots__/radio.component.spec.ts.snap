// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackfestRadioComponent Host Views should display only one option 1`] = `
<div
  class="hackfest-radio"
>
  <hackfest-radio-option
    value="1"
  >
    <div
      class="hackfest-radio"
    >
      <input
        aria-checked="false"
        class="cdk-visually-hidden"
        id="hackfest-radio-option-0"
        name="undefined"
        type="radio"
        value="1"
      />
      <label
        for="hackfest-radio-option-0"
      >
        1
        <svg
          viewBox="0 0 100 100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M27.8,18.6c-3.7,5.1-7.3,10.1-11,15.2c13.7-6.9,27.5-13.9,42.4-17.6c-4.5,5.5-10.4,9.7-15.7,14.5
c-4.4,4-8.5,8.4-12.5,12.8C25,50,19,56.5,14.1,63.9c10.2-6.1,20.4-12.2,30.6-18.4c3.1-1.9,6.2-3.7,9.4-5.2c3.2-1.4,6.5-2.5,9.7-3.7
c7.5-2.8,14.7-6.5,21.8-10.2c-4.2,3.7-9.4,6.1-13.9,9.4c-3.1,2.3-5.8,4.9-8.8,7.4c-3.5,2.9-7.3,5.4-10.7,8.4
c-4.5,4.2-8,9.3-11.6,14.2c-4.6,6.1-9.6,11.9-14.3,17.9c9.7-8.4,17-19.5,27.5-26.9c7.1-5,15.3-8.2,23.6-10.6
c2.9-0.8,5.8-1.6,8.4-3.1c-3.5,4.9-8.2,8.7-12.8,12.6c-7.6,6.7-14.5,14-20.7,22C62.7,74.1,71,66.1,81.4,63.2"
          />
        </svg>
      </label>
    </div>
  </hackfest-radio-option>
</div>
`;

exports[`HackfestRadioComponent Host Views should display radio with options 1`] = `
<div
  class="hackfest-radio"
>
  
  <hackfest-radio-option>
    <div
      class="hackfest-radio"
    >
      <input
        aria-checked="false"
        class="cdk-visually-hidden"
        id="hackfest-radio-option-1"
        name="undefined"
        type="radio"
        value="undefined"
      />
      <label
        for="hackfest-radio-option-1"
      >
        
        <svg
          viewBox="0 0 100 100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M27.8,18.6c-3.7,5.1-7.3,10.1-11,15.2c13.7-6.9,27.5-13.9,42.4-17.6c-4.5,5.5-10.4,9.7-15.7,14.5
c-4.4,4-8.5,8.4-12.5,12.8C25,50,19,56.5,14.1,63.9c10.2-6.1,20.4-12.2,30.6-18.4c3.1-1.9,6.2-3.7,9.4-5.2c3.2-1.4,6.5-2.5,9.7-3.7
c7.5-2.8,14.7-6.5,21.8-10.2c-4.2,3.7-9.4,6.1-13.9,9.4c-3.1,2.3-5.8,4.9-8.8,7.4c-3.5,2.9-7.3,5.4-10.7,8.4
c-4.5,4.2-8,9.3-11.6,14.2c-4.6,6.1-9.6,11.9-14.3,17.9c9.7-8.4,17-19.5,27.5-26.9c7.1-5,15.3-8.2,23.6-10.6
c2.9-0.8,5.8-1.6,8.4-3.1c-3.5,4.9-8.2,8.7-12.8,12.6c-7.6,6.7-14.5,14-20.7,22C62.7,74.1,71,66.1,81.4,63.2"
          />
        </svg>
      </label>
    </div>
  </hackfest-radio-option>
  <hackfest-radio-option>
    <div
      class="hackfest-radio"
    >
      <input
        aria-checked="false"
        class="cdk-visually-hidden"
        id="hackfest-radio-option-2"
        name="undefined"
        type="radio"
        value="undefined"
      />
      <label
        for="hackfest-radio-option-2"
      >
        
        <svg
          viewBox="0 0 100 100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M27.8,18.6c-3.7,5.1-7.3,10.1-11,15.2c13.7-6.9,27.5-13.9,42.4-17.6c-4.5,5.5-10.4,9.7-15.7,14.5
c-4.4,4-8.5,8.4-12.5,12.8C25,50,19,56.5,14.1,63.9c10.2-6.1,20.4-12.2,30.6-18.4c3.1-1.9,6.2-3.7,9.4-5.2c3.2-1.4,6.5-2.5,9.7-3.7
c7.5-2.8,14.7-6.5,21.8-10.2c-4.2,3.7-9.4,6.1-13.9,9.4c-3.1,2.3-5.8,4.9-8.8,7.4c-3.5,2.9-7.3,5.4-10.7,8.4
c-4.5,4.2-8,9.3-11.6,14.2c-4.6,6.1-9.6,11.9-14.3,17.9c9.7-8.4,17-19.5,27.5-26.9c7.1-5,15.3-8.2,23.6-10.6
c2.9-0.8,5.8-1.6,8.4-3.1c-3.5,4.9-8.2,8.7-12.8,12.6c-7.6,6.7-14.5,14-20.7,22C62.7,74.1,71,66.1,81.4,63.2"
          />
        </svg>
      </label>
    </div>
  </hackfest-radio-option>
</div>
`;
